### DISCUSSION TITLE
v24.12.16

### DISCUSSION THREAD

**Discussion by davidrapan** [davidrapan/ha-solarman#327] [Announcements]
## What's Changed
* [fix: Missing extra state attributes](https://github.com/davidrapan/ha-solarman/commit/70a3f764d4a5212bea3a9daa7ad25415293c1713)

**Full Changelog**: https://github.com/davidrapan/ha-solarman/compare/v24.12.14...v24.12.16

<hr /><em>This discussion was created from the release <a href='https://github.com/davidrapan/ha-solarman/releases/tag/v24.12.16'>v24.12.16</a>.</em>

**Comment by sofkaski** [comment]
I installed `v24.12.16` and did again some silk glove testing. Almost everything is working, but I got some dirt to my gloves from time-of-use (TOU) stuff. 

I have not paid too much attention to this since in my configuration TOU is on every weekday. Now one of my dashboard cards that shows a summary of TOU programs broke. It used to check if TOU is enabled or not by comparing the state of `select.inverter_time_of_use` with `Enabled`. Now it broke,because the state was `Week` (decimal 255,  binary 11111111).

In the inverter UI there are separate 'Enable' tick-box  and tick-boxes for each weekday.
If I ticked Sunday off, the result in Solarman was `Disabled` (decimal 127, binary 01111111). And if I ticked Saturday off the result was `Disabled` (decimal 191, binary 10111111).

Then I noticed that there was a new fix b4059b0584 and I installed the latest from main (a08ed6fa5). The behavior changed so that off- ticked Sunday results in `Week-Sunday` and ticking off Sunday and Friday results in `Disabled` (decimal 95, binary 01011111).
So my guess about Deye register  is that the lowest bit is enable/disable for the whole TOU and then a bit per each weekday. Sunday being the highest bit. 

I think as concluded in #319 the proper solution here would be a multi-select selector (which does not exist in HA atm).

**Reply by davidrapan** [reply]
Yes i had to somehow differentiate enabled for whole week and cases which are not in the options but still means "enabled".

So you are saying that the fall back for any value starting with bit 0 set to one is not working correctly?

**Reply by sofkaski** [reply]
So it seems. I started doubting my notes, but looking from history it really seems that the value has been `Disabled` around time where I ticked Sunday and Friday off (~17:15).

![image](https://github.com/user-attachments/assets/5bd698f4-8609-4a1c-b3be-65b7e52b3828)

**Reply by davidrapan** [reply]
[fix: Enum and select bit lookup w/ single mode](https://github.com/davidrapan/ha-solarman/commit/24ce03f8b510302650b8090453d85e2e381de6c9)
